README.md
# readme-inspector [![NPM version][npm-image]][npm-url] [![GitHub release][github-release-image]][github-release-url]

> <img align="middle" alt="markdown" height="50" width="50"  src="https://cdnjs.cloudflare.com/ajax/libs/octicons/4.4.0/svg/markdown.svg"> Inspect GitHub (Enterprise) repositories for the presence and quality of READMEs.

[![The MIT License][license-image]][license-url]
[![FOSSA Status][fossa-image]][fossa-url]
[![NSP Status][nsp-image]][nsp-url]
[![Known Vulnerabilities][vulnerabilities-image]][vulnerabilities-url]<br>
[![Dependency Status][daviddm-image]][daviddm-url]
[![Development Dependency Status][daviddm-dev-image]][daviddm-dev-url]<br>
[![MacOS and Ubuntu build statuses][travis-image]][travis-url]
[![Windows build status][appveyor-image]][appveyor-url]
[![Coverage percentage][codacy-coverage-image]][codacy-url]
[![Codacy code quality][codacy-image]][codacy-url]
![Maintenance][maintenance-image]<br>
[![NPMS score][npms-image]][npms-url]
[![NPM downloads per month][npm-downloads-month]][npm-url]

## Table of contents

<!-- 🚫 AUTO-GENERATED-CONTENT:START (TOC:excludeText=Table of contents) -->
- [1. Installation](#1-installation)
- [2. Configuration](#2-configuration)
  * [2.1. Setting environment variables](#21-setting-environment-variables)
- [2. Usage](#2-usage)
- [3. API](#3-api)
  * [3.1. `authenticate({token, type, key})`](#31-authenticatetoken-type-key)
    + [3.1.1. Parameters](#311-parameters)
    + [3.1.2. Returns `void`](#312-returns-void)
    + [3.1.3. Example](#313-example)
  * [3.2. `check({ower, repo, ref})`](#32-checkower-repo-ref)
    + [3.2.1. Parameters](#321-parameters)
    + [3.2.2. Returns `Promise`](#322-returns-promise)
    + [3.2.3. Examples](#323-examples)
  * [3.3. `getInfo({owner, repo, ref})`](#33-getinfoowner-repo-ref)
    + [3.3.1. Parameters](#331-parameters)
    + [3.3.2. Returns `Promise`](#332-returns-promise)
    + [3.3.3. Examples](#333-examples)
  * [3.4. `getAppraisal(url)`](#34-getappraisalurl)
  * [3.5. `ReadmeAppraisal`](#35-readmeappraisal)
    + [3.5.1. `for(url): Promise`](#351-forurl-promise)
      - [3.5.1.1. Parameters](#3511-parameters)
      - [3.5.1.2. Returns `Promise`](#3512-returns-promise)
      - [3.5.1.3. Examples](#3513-examples)
- [4. Version](#4-version)
- [5. Contributing](#5-contributing)
- [6. License](#6-license)
<!-- 🚫 AUTO-GENERATED-CONTENT:END -->

## 1. Installation

`readme-inspector` is written in JavaScript (CommonJS) for [Node.js ![External link][octicon-link-external]](https://nodejs.org/) versions 7.6.0 or higher (for `async/await` support).

```bash
$ npm install --save readme-inspector
```

## 2. Configuration

The `commonality/readme-inspector` module combines the mediator, proxy, and factory design patterns to simplify:

* README detection with the `readmeInfo` object, and
* Quality assessment with the `readmeInfo.appraisal` object.

### 2.1. Setting environment variables

`readme-inspector` invokes Web services to return information. These services use `.env` variables
that require configuration:

```properties
# ENV VARS defaults for readme-inspector:

# ReadmeAppraisal REST API
# 💼 Modify this if you are using a company hosted installation.
API_ENDPOINT_README_SCORE="http://readme-score-api.herokuapp.com/score.json?url=&human_breakdown=false&force=false"

# Bitbucket REST API v1.0 and v2.0 base url.
# 💼 Modify this if you're using on-premise,
#    company-hosted Bitbucket servers.
BITBUCKET_API_BASE_URL="https://api.bitbucket.org"

## Google Analytics trackingCode
GA_README_INSPECTOR="UA-117338111-1"

# GitHub REST API v3 baseUrl.
# 💼 Modify this if you're using GitHub Enterprise, e.g.,
# GITHUB_API_BASE_URL="https://evilcorp.github.com/v3
GITHUB_API_BASE_URL="https://api.github.com"

# readme-score-api base url.
# 💼 Modify this if you've installed it behind a company firewall.
README_SCORE_API_BASE_URL="http://readme-score-api.herokuapp.com"
```

> ![light-bulb][octicon-light-bulb] **To avoid rate-limiting**, you should [create a personal access token ![External link][octicon-link-external]](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/) and save your personal access token in an environment variable called `GH_TOKEN`.

---

<details><summary>Click here for detailed <samp>.env</samp> variable initialization instructions.</summary><p>

> [![info][octicon-info] View **dotenv-extended**'s README ![External link][octicon-link-external]](https://github.com/keithmorris/node-dotenv-extended#readme) for detailed `.env` variable set up instructions.

<h4><img align="bottom" alt="file" src="https://cdnjs.cloudflare.com/ajax/libs/octicons/4.4.0/svg/file.svg"> <samp>.env.schema</samp></h4>

Defines a schema of what variables should be defined in the combination of
<samp>.env</samp> and <samp>.env.defaults</samp>.

<!-- AUTO-GENERATED-CONTENT:START (CODE:src=./.env.schema&syntax=properties) -->
<!-- The below code snippet is automatically added from ./.env.schema -->
```properties
# .env.schema, committed to repo

## See https://github.com/keithmorris/node-dotenv-extended/#readme
## ⛔️
## 🚫  DO NOT COMMIT YOUR ACTUAL .env file to version control.
## 🚫  It should only include environment-specific values such
## 🚫  as database passwords or API keys.
## 🚫  Your production database should have a different password
## 🚫  than your development database.

# ENV VARS required for readme-inspector
## Add values to these ENV VARs and save to
## {your-project-root-directory}/.env, e.g.,
## if your root project directory is
##
## $ more-bort-license-plates/
##
## then save your .env file to
##
## $ more-bort-license-plates/.env

# 🔹 OPTIONAL env vars (set in .env.defaults):
API_ENDPOINT_README_SCORE=
BITBUCKET_API_BASE_URL=
GA_README_INSPECTOR=
GITHUB_API_BASE_URL=
README_SCORE_API_BASE_URL=

# 🔸 RECOMMENDED vars (to extend API rate limits)
BITBUCKET_ACCESS_TOKEN=
BITBUCKET_AUTH_PASSWORD=
BITBUCKET_AUTH_USERNAME=
GH_TOKEN=
GITHUB_ACCESS_TOKEN=
```
<!-- AUTO-GENERATED-CONTENT:END -->

<h4><img align="bottom" alt="file" src="https://cdnjs.cloudflare.com/ajax/libs/octicons/4.4.0/svg/file.svg"> <samp>.env.defaults</samp></h4>

<samp>.env.defaults</samp> provides common configuration defaults across all
environments (commited to source control). This contains overall app
configuration values that would be common across environments. The
<samp>.env.defaults</samp> file is loaded first; then the <samp>.env</samp>
file is loaded and will overwrite any values from the <samp>.env.defaults</samp>
file.

<!-- AUTO-GENERATED-CONTENT:START (CODE:src=./.env.defaults&syntax=properties) -->
<!-- The below code snippet is automatically added from ./.env.defaults -->
```properties
# .env.defaults, committed to repo

## See https://github.com/keithmorris/node-dotenv-extended/#readme
## ⛔️
## 🚫  DO NOT COMMIT YOUR ACTUAL .env file to version control.
## 🚫  It should only include environment-specific values such
## 🚫  as database passwords or API keys.
## 🚫  Your production database should have a different password
## 🚫  than your development database.

# ENV VARS defaults for readme-inspector:

# ReadmeAppraisal REST API
# 💼 Modify this if you are using a company hosted installation.
API_ENDPOINT_README_SCORE="http://readme-score-api.herokuapp.com/score.json?url=&human_breakdown=false&force=false"

# Bitbucket REST API v1.0 and v2.0 base url.
# 💼 Modify this if you're using on-premise, company-hosted Bitbucket servers.
BITBUCKET_API_BASE_URL="https://api.bitbucket.org"

## Google Analytics trackingCode
GA_README_INSPECTOR="UA-117338111-1"

# GitHub REST API v3 baseUrl.
# 💼 Modify this if you're using GitHub Enterprise.
GITHUB_API_BASE_URL="https://api.github.com"

# readme-score-api base url.
# 💼 Modify this if you've installed it behind a company firewall.
README_SCORE_API_BASE_URL="http://readme-score-api.herokuapp.com"
```
<!-- AUTO-GENERATED-CONTENT:END -->

<h4><img align="bottom" alt="file" src="https://cdnjs.cloudflare.com/ajax/libs/octicons/4.4.0/svg/file.svg"> <samp>.env</samp></h4>

The environment-specific file (not committed to source control).
This file will have sensitive information such as usernames, passwords,
api keys, etc. These would be specific to each environment and **should
not be committed to source control**.

<!-- AUTO-GENERATED-CONTENT:START (CODE:src=./.env&syntax=properties) -->
<!-- The below code snippet is automatically added from ./.env -->
```properties
# BITBUCKET_ACCESS_TOKEN=
# BITBUCKET_API_BASE_URL=
# BITBUCKET_AUTH_PASSWORD=
# BITBUCKET_AUTH_USERNAME=
# GA_README_INSPECTOR="UA-117338111-1"
# GITHUB_API_BASE_URL=
# README_SCORE_API_BASE_URL=
API_ENDPOINT_README_SCORE="http://readme-score-api.herokuapp.com/score.json?url=&human_breakdown=false&force=false"
GH_TOKEN=$(echo $GH_TOKEN)
GITHUB_ACCESS_TOKEN=$(echo $GH_TOKEN)
```
<!-- AUTO-GENERATED-CONTENT:END -->

</pre></details>

---

## 2. Usage

> [![beaker][octicon-beaker] Test `readme-inspector` in your Web browser ![link-external][octicon-link-external]][runkit-readme-inspector-url].

<!-- AUTO-GENERATED-CONTENT:START (CODE:src=./.github/assets/examples/usage.js) -->
<!-- The below code snippet is automatically added from ./.github/assets/examples/usage.js -->
```js
// Load all .env variables before anything else.

const dotenvExtended = require('dotenv-extended')
const envConfig = dotenvExtended.config()

// Import readme-inspector.

const readmeInspector = require('readme-inspector')

// Recommended: authenticate to avoid rate limts.

readmeInspector.authenticate({
  token: envConfig.GH_TOKEN,
  type: 'oauth'
})

// Verify that the repository with the slug
// gregswindle/github-resource-converter
// 1. Has a README, and
// 2. Score the README for quality.

const info = await readmeInspector.check({
  owner: 'gregswindle',
  repo:  'github-resource-converter'
})

// Display the resulting readmeInfo as a
// JSON string.

const WHITESPACE = 2
console.log(JSON.stringify(results, null, WHITESPACE))
// =>
/*
{
  "appraisal": {
    "breakdown": {
      "cumulativeCodeBlockLength": 0,
      "hasLists": 0,
      "lowCodeBlockPenalty": 0,
      "numberOfCodeBlocks": 0,
      "numberOfGifs": 0,
      "numberOfImages": 0,
      "numberOfNonCodeSections": 0
    },
    "error": null,
    "score": 0,
    "url": null
  },
  "error": null,
  "isPresent": true,
  "value": {
    "name": "README.md",
    "path": "README.md",
    "sha": "4769744aad57ff3e9aac2df603795c4d10fcdc31",
    "size": 36877,
    "url": "https://api.github.com/repos/commonality/readme-inspector/contents/README.md?ref=master",
    "html_url": "https://github.com/commonality/readme-inspector/blob/master/README.md",
    "git_url": "https://api.github.com/repos/commonality/readme-inspector/git/blobs/4769744aad57ff3e9aac2df603795c4d10fcdc31",
    "download_url": "https://raw.githubusercontent.com/commonality/readme-inspector/master/README.md",
    "type": "file",
    "content": "{base64-encoding-of-readme-markdown}",
    "encoding": "base64",
    "_links": {
      "self": "https://api.github.com/repos/commonality/readme-inspector/contents/README.md?ref=master",
      "git": "https://api.github.com/repos/commonality/readme-inspector/git/blobs/4769744aad57ff3e9aac2df603795c4d10fcdc31",
      "html": "https://github.com/commonality/readme-inspector/blob/master/README.md"
    }
  }
}
*/
```
<!-- AUTO-GENERATED-CONTENT:END -->

## 3. API

> [![beaker][octicon-beaker] Test `readme-inspector` in your Web browser ![link-external][octicon-link-external]][runkit-readme-inspector-url].
>
> [![gear][octicon-gear] View the full API docs for details ![link-external][octicon-link-external]][api-docs-url].

The `readmeInspector` module detects whether or not a README document exists at the root of a GitHub or GitHub Enterprise repository. If a README exists, it can evaluate the README's quality and provide a numerical score from 0 to 100, where 0 is the lowest quality and 100 is the highest.

### 3.1. `authenticate({token, type, key})`

> ![Info][octicon-info] Most GitHub API calls don't require authentication. Rules of thumb:
>
> 1.  If you can see the information by visiting the site without being logged in, you don't have to be authenticated to retrieve the same information through the API.
> 1.  If you want to change data, you have to be authenticated.
>
> Note: authenticate is synchronous because it only sets the credentials for the following requests.
>
> octokit/rest.js. (2018). GitHub. Retrieved 21 March 2018, from <https://github.com/octokit/rest.js#authentication> ![link-external][octicon-link-external]

#### 3.1.1. Parameters

| Name     | Type          | Description                                 |
| :------- | :------------ | :------------------------------------------ |
| key      | String        | `type=oauth` Client identifier              |
| token    | String        | `type=[integration|token]` Unique value     |
| type     | Enum.<String> | `basic`, `oauth`, `token`, or `integration` |
| username | String        | `type=basic` Basic authentication username  |
| password | String        | `type=basic` Basic authentication password  |

#### 3.1.2. Returns `void`

`authenticate` does not return a value.

#### 3.1.3. Example

> ```javascript
> // Token (https://github.com/settings/tokens)
> // Load your GH_TOKEN or GITHUB_ACCESS_TOKEN from
> // environment variables:
> const dotenvExtended = require('dotenv-extended')
> const envConfig = dotenvExtended.config()
>
> const readmeInspector = require('readme-inspector')
>
> readmeInspector.authenticate({
>   token: envConfig.GH_TOKEN,
>   type: 'token'
> })
> ```

### 3.2. `check({ower, repo, ref})`

A convenience method that

* Attempts to <samp>GET</samp> a repository's root-level README, and, if found,
* Scores the README.

![GET][rest-get-img]

```http
/repos/:owner/:repo/readme
```

If a README is found, then the `readme-score-api` is invoked:

![GET][rest-get-img]

```http
/score.json?url=:owner/:repo
```

#### 3.2.1. Parameters

<table>
  <thead>
    <tr>
    <th style="width: 30%">Field</th>
      <th style="width: 10%">Type</th>
      <th style="width: 60%">Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><samp>owner</samp></td>
        <td>
          String
        </td>
      <td> </td>
    </tr>
    <tr>
      <td><samp>repo</samp></td>
        <td>
          String
        </td>
      <td> </td>
    </tr>
    <tr>
      <td><samp>ref</samp> <img align="right" alt="optional" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16" height="22" width="60"></td>
        <td>
          String
        </td>
      <td>The name of the commit/branch/tag. Default: the repository’s default branch (usually master).</td>
    </tr>
  </tbody>
</table>

#### 3.2.2. Returns `Promise<ReadmeInfo>`

`ReadmeInfo's` interface (as a `NullObject`):

```js
{
  'err': null,
  'isPresent': null,
  'appraisal': {
    'breakdown': {
      'cumulativeCodeBlockLength': 0,
      'hasLists': 0,
      'lowCodeBlockPenalty': 0,
      'numberOfCodeBlocks': 0,
      'numberOfGifs': 0,
      'numberOfImages': 0,
      'numberOfNonCodeSections': 0
    },
    'err': null,
    'score': 0,
    'url': null
  },
  'value': null
}
```

#### 3.2.3. Examples

* _async/await:_

  > ```js
  > const readmeInfo = await readmeInspector.check({
  >   owner: 'commonality',
  >   ref: 'GH-1-a-topic-branch',
  >   repo: 'readme-inspector'
  > })
  > ```

* _Promise:_

  > ```js
  > readmeInspector
  >   .check({
  >     owner: 'commonality',
  >     ref: 'GH-1-a-topic-branch',
  >     repo: 'readme-inspector'
  >   })
  >   .then(readmeInfo => {})
  >   .catch(err => {})
  > ```

### 3.3. `getInfo({owner, repo, ref})`

Retrieves README information _without_ any `AppraisalData`.

![GET][rest-get-img]

```http
/repos/:owner/:repo/readme
```

#### 3.3.1. Parameters

<table>
  <thead>
    <tr>
    <th style="width: 30%">Field</th>
      <th style="width: 10%">Type</th>
      <th style="width: 60%">Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><samp>owner</samp></td>
        <td>
          String
        </td>
      <td> </td>
    </tr>
    <tr>
      <td><samp>repo</samp></td>
        <td>
          String
        </td>
      <td> </td>
    </tr>
    <tr>
      <td><samp>ref</samp> <img align="right" alt="optional" src="https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16" height="22" width="60"></td>
        <td>
          String
        </td>
      <td>The name of the commit/branch/tag. Default: the repository’s default branch (usually master).</td>
    </tr>
  </tbody>
</table>

#### 3.3.2. Returns `Promise<ReadmeInfo>`

`ReadmeInfo's` interface (as a `NullObject`):

```js
{
  'err': null,
  'isPresent': null,
  'appraisal': {
    'breakdown': {
      'cumulativeCodeBlockLength': 0,
      'hasLists': 0,
      'lowCodeBlockPenalty': 0,
      'numberOfCodeBlocks': 0,
      'numberOfGifs': 0,
      'numberOfImages': 0,
      'numberOfNonCodeSections': 0
    },
    'err': null,
    'score': 0,
    'url': null
  },
  'value': null
}
```

#### 3.3.3. Examples

* _async/await:_

  > ```js
  > const readmeInfo = await readmeInspector.getInfo({
  >   owner: 'commonality',
  >   ref: 'GH-1-a-topic-branch',
  >   repo: 'readme-inspector'
  > })
  > ```

* _Promise:_

  > ```js
  > readmeInspector
  >   .getInfo({
  >     owner: 'commonality',
  >     ref: 'GH-1-a-topic-branch',
  >     repo: 'readme-inspector'
  >   })
  >   .then(readmeInfo => readmeInfo)
  >   .catch(err => {
  >     console.error(err)
  >   })
  > ```

### 3.4. `getAppraisal(url)`

A convenience wrapper that calls the `ReadmeAppraisal.prototype.for` method.

### 3.5. `ReadmeAppraisal`

`ReadmeAppraisal` is an API proxy for [@clayallsopp ![External link][octicon-link-external]](https://github.com/clayallsopp)'s [`readme-score-api` ![External link][octicon-link-external]](https://github.com/clayallsopp/readme-score-api).

> ![quote][octicon-quote] ScoreMe gives you a numerical score from 0 to 100 for your Github-style README. The intention is to measure complexity, which is a generally correlated with quality.
>
> It won't measure if one README is absolutely better than another, but it will give you a good idea if the README is high-quality, needs more work, or somewhere inbetween.
>
> ScoreMe. (2018). Clayallsopp.github.io. Retrieved 10 April 2018, from <http://clayallsopp.github.io/readme-score/>

#### 3.5.1. `for(url): Promise<AppraisalData>`

Evaluate the README at the root of a GitHub repository.

##### 3.5.1.1. Parameters

| Name | Type   | Description                                                      |
| :--- | :----- | :--------------------------------------------------------------- |
| url  | String | The URL, or slug of the repository to be evaluated for a README. |

##### 3.5.1.2. Returns `Promise<AppraisalData>`

* `AppraisalData` as a `NullObject` (see <samp>[lib/appraisal-data](lib/appraisal-data.js)</samp>):<br>

  > ```js
  > {
  >   breakdown: {
  >     cumulativeCodeBlockLength: 0
  >     hasLists: 0
  >     lowCodeBlockPenalty: 0
  >     numberOfCodeBlocks: 0
  >     numberOfGifs: 0
  >     numberOfImages: 0
  >     numberOfNonCodeSections: 0
  >   },
  >   err: null,
  >   score: 0
  >   url: null
  > }
  > ```

##### 3.5.1.3. Examples

* _URL:_

  > ```js
  > const { ReadmeAppraisal } = require('readme-inspector')
  > const readmeAppraisal = new ReadmeAppraisal()
  > const url = 'https://github.com/gregswindle/github-resource-converter'
  >
  > const appraisal = readmeAppraisal.for(url)
  > /** =>
  >  * {
  >  *   breakdown: {
  >  *    cumulativeCodeBlockLength: 10
  >  *    hasLists: 10
  >  *    lowCodeBlockPenalty: 0
  >  *    numberOfCodeBlocks: 40
  >  *    numberOfGifs: 0
  >  *    numberOfImages: 15
  >  *    numberOfNonCodeSections: 30
  >  *  },
  >  *  err: null,
  >  *  score: 100
  >  *  url: 'https://github.com/gregswindle/github-resource-converter'
  >  * }
  >  */
  > ```

* _Repository slug:_

  > ```js
  > const { ReadmeAppraisal } = require('readme-inspector')
  > const readmeAppraisal = new ReadmeAppraisal()
  > const url = 'gregswindle/github-resource-converter'
  >
  > const appraisal = readmeAppraisal.for(url)
  > ```

## 4. Version

[![NPM version][npm-image]][npm-url]

View the [Change Log](CHANGELOG.md) and [Releases](https://github.com/commonality/readme-inspector/releases) for details.

## 5. Contributing

[![PRs Welcome][makeapullrequest-image] ![External link][octicon-link-external]][makeapullrequest-url] We welcome contributions with GitHub **issues** and **pull requests**.

---

[![Request a feature][issues-new-feat-image]][issues-new-feat-url]
[![Report a defect][issues-new-defect-image]][issues-new-defect-url]

[![Read the CONTRIBUTING guidelines][contributing-image]][contributing-url]

---

Before embarking on a significant change, please follow these guidelines:

1.  **[Create an issue][issues-url]**&mdash;e.g., a [defect ("bug") report][issues-new-defect-url] or a [feature request][issues-new-feat-url]&mdash;to propose changes.

    _Exceptions:_

    > If you're working on documentation and fixing something simple like a typo or an easy bug, go ahead and make a pull request.

1.  **[Follow the CONTRIBUTING guidelines][contributing-url].**

    _Why:_

    > Standards and guidelines make communication easier. If you're willing and able to program&mdash;or want to learn how&mdash; following the guidelines will increase the likelihood of having your changes added to `readme-inspector`.

1.  **[Read the Code of Conduct][code-of-conduct-url].**<br>

1.  **[Make a pull request][pr-url]** when you're ready for other to review your changes (or you get stuck somewhere).

    _Never created a pull request?_

    > No problem: [this free online training ![External link][octicon-link-external]][makeapullrequest-url] covers most of the conventions in the [CONTRIBUTING guidelines][contributing-url].)

## 6. License

[MIT ![link-external][octicon-link-external]][license-url] © [commonality ![link-external][octicon-link-external]](https://github.com/commonality)

[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fcommonality%2Freadme-inspector.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2Fcommonality%2Freadme-inspector?ref=badge_large)

---

[![Greenkeeper badge](https://badges.greenkeeper.io/commonality/readme-inspector.svg)](https://greenkeeper.io/)
[![Readme ReadmeAppraisal](http://readme-score-api.herokuapp.com/score.svg?url=https://github.com/commonality/readme-inspector)](http://clayallsopp.github.io/readme-score?url=https://github.com/commonality/readme-inspector)

<!-- ⛔️ Do not remove this comment or anything under it ⛔️ -->

<!-- 🔗 link references 🔗 -->

[api-docs-url]: https://commonality.github.io/readme-inspector/api/readme-inspector/2.0.1/ReadmeAppraisal.html
[bunyan-format-url]: https://github.com/thlorenz/bunyan-format/#readme
[node-bunyan-url]: https://github.com/trentm/node-bunyan/#readme
[optional-param-img]: https://fakeimg.pl/60x22/757575/FFF/?text=optional&font_size=16
[rest-get-img]: https://fakeimg.pl/40x30/0e8a16/FFF/?text=GET&font_size=20
[runkit-readme-inspector-url]: https://runkit.com/gregswindle/5acc09bde794d70011a136e5
[toc]: #table-of-contents

<!-- 🔗 ci services 🔗 -->

[appveyor-image]: https://img.shields.io/appveyor/ci/gregswindle/readme-inspector.svg?style=flat-square&logo=appveyor&label=windows%20build
[appveyor-url]: https://ci.appveyor.com/project/gregswindle/readme-inspector
[codacy-coverage-image]: https://img.shields.io/codacy/coverage/21f517a2d5bf4304895f40c5cbb596c4.svg?style=flat-square
[codacy-image]: https://img.shields.io/codacy/grade/21f517a2d5bf4304895f40c5cbb596c4.svg?style=flat-square
[codacy-url]: https://www.codacy.com/app/greg_7/readme-inspector?utm_source=github.com&utm_medium=referral&utm_content=commonality/readme-inspector&utm_campaign=Badge_Grade
[coveralls-image]: https://img.shields.io/coveralls/github/commonality/readme-inspector/master.svg
[coveralls-url]: https://coveralls.io/r/commonality/readme-inspector
[daviddm-dev-image]: https://david-dm.org/commonality/readme-inspector/dev-status.svg?theme=shields.io&style=flat-square
[daviddm-dev-url]: https://david-dm.org/commonality/readme-inspector?type=dev
[daviddm-image]: https://david-dm.org/commonality/readme-inspector.svg?theme=shields.io&style=flat-square

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="2.0.7"></a>

## [2.0.7](https://github.com/commonality/readme-inspector/compare/v2.0.6...v2.0.7) (2018-05-01)

### Dependency Updates

* **jest-runner-eslint:** remove unused library ([a68be53](https://github.com/commonality/readme-inspector/commit/a68be53))

---

<a name="2.0.6"></a>

## [2.0.6](https://github.com/commonality/readme-inspector/compare/v2.0.5...v2.0.6) (2018-05-01)

### Dependency Updates

* **ajv-keywords:** update to version 3.2.0 ([8be7236](https://github.com/commonality/readme-inspector/commit/8be7236))

---

<a name="2.0.5"></a>

## [2.0.5](https://github.com/commonality/readme-inspector/compare/v2.0.4...v2.0.5) (2018-04-27)

### Dependency Updates

* **codacy-coverage:** update to v3.0.0 ([3892fc5](https://github.com/commonality/readme-inspector/commit/3892fc5))
* **codacy-coverage:** update to version 3.0.0 ([6990e17](https://github.com/commonality/readme-inspector/commit/6990e17))
* **lint-staged:** update to version 7.0.5 ([5c257cd](https://github.com/commonality/readme-inspector/commit/5c257cd))

---

<a name="2.0.2"></a>

## [2.0.2](https://github.com/commonality/readme-inspector/compare/v2.0.3...v2.0.2) (2018-04-26)

### Dependency Updates

* **lint-staged:** update to version 7.0.5 ([5c257cd](https://github.com/commonality/readme-inspector/commit/5c257cd))

---

<a name="2.0.1"></a>

## [2.0.1](https://github.com/commonality/readme-inspector/compare/v2.0.0...v2.0.1) (2018-04-24)

### Bug Fixes

* **config:** move .env files ([2ff8905](https://github.com/commonality/readme-inspector/commit/2ff8905)), closes [#23](https://github.com/commonality/readme-inspector/issues/23)

---

<a name="2.0.0"></a>

## [2.0.0](https://github.com/commonality/readme-inspector/compare/v1.0.2...v2.0.0) (2018-04-21)

### Dependency Updates

* **@semantic-release/git:** update to version 4.0.2 ([ad96ea5](https://github.com/commonality/readme-inspector/commit/ad96ea5))

### Bug Fixes

* **score:** verify appraisal initialization ([fd87a8b](https://github.com/commonality/readme-inspector/commit/fd87a8b)), closes [#18](https://github.com/commonality/readme-inspector/issues/18) [#21](https://github.com/commonality/readme-inspector/issues/21)

### BREAKING CHANGES

**readme-score:** `ReadmeScore` is now `ReadmeAppraisal`, with two method signatures changes and one property name change.

#### Objects & Properties

`readmeInspector.ReadmeScore` - refactor:rename `ReadmeScore` → `ReadmeAppraisal`, which is now a class.

#### Functions

1.  `readmeInspector.getReadmeInfo` → `readmeInspector.getInfo`

    > * Refactor:rename function
    >
    > * Refactor function signature: now expects a `params` object literal

2.  `readmeInspector.getReadmeScore` → `readmeInspector.getAppraisal`

    > Refactor:rename function; signature remains the same as in 1.x.x,
    > however.

3.  `readmeInspector.check`'s signature now expects a `params` object literal,
    which it passes to `getInfo`.

#### API tests

Add API (integration) tests.

---

<a name="1.0.2"></a>

## [1.0.2](https://github.com/commonality/readme-inspector/compare/v1.0.1...v1.0.2) (2018-04-17)

### Bug Fixes

* **null-readme-score:** add url to search ([49d4065](https://github.com/commonality/readme-inspector/commit/49d4065)), closes [#18](https://github.com/commonality/readme-inspector/issues/18) [#19](https://github.com/commonality/readme-inspector/issues/19)

  > [`ReadmeScore.for`](https://github.com/commonality/readme-inspector/blob/master/lib/readme-score.js#L60)&mdash;and consequently, [`readmeInspector.getReadmeScore`](https://github.com/commonality/readme-inspector/blob/master/lib/index.js#L150)&mdash;returning `nullScoreData`, i.e., scores for READMEs that should have high marks.
  >
  > Refactor `readmeScoreApiClientOptions` with a single property called `apiEndpoint: URL` in order to set the `url` query parameter explicitly, and call `apiEndpoint.toString()` for a serialized URL string.

---

<a name="1.0.1"></a>

## [1.0.1](https://github.com/commonality/readme-inspector/compare/v1.0.0...v1.0.1) (2018-04-16)

### Bug Fixes

* **config:** add default .env var values ([7a71637](https://github.com/commonality/readme-inspector/commit/7a71637)), closes [#11](https://github.com/commonality/readme-inspector/issues/11) [#10](https://github.com/commonality/readme-inspector/issues/10).

  Replaced [`dotenv`][dotenv-url] with [`dotenv-extended`][dotenv-extended-url] in order to load default values.

---

<a name="1.0.0"></a>

## 1.0.0 (2018-04-12)

### Features

* #### module:readme-inspector

  Verify the existence—and assess the quality—of README files

  > ![quote][octicon-quote] READMEs do more than explain how to use your project. They also
  > explain why your project matters, and what your users can do with it.
  >
  > In your README, try to answer the following questions:
  >
  > * What does this project do?
  > * Why is this project useful?
  > * How do I get started?
  > * Where can I get more help, if I need it?
  >
  > You can use your README to answer other questions, like how you handle
  > contributions, what the goals of the project are, and information about
  > licenses and attribution. If you don’t want to accept contributions, or
  > your project is not yet ready for production, write this information down. <sup><a href="#ref-1" title="View reference.">[1]</a></sup>

* #### Public API

  _Methods_

  > * `authenticate` - Sets GitHub credentials for all subsequent requests.
  > * `check` - Attempts to GET and assess a README at a repo-root directory.
  > * `getReadmeInfo` - Attempt to GET a README without assessing it.
  > * `getReadmeScore` - Assess the quality of a README.

  _Properties_

  > * `ReadmeScore` - An API proxy wrapper for the readme-score-api.
  > * `api` - A configurable Octokit instance.


lib/readme-info.js
const appraisalData = require('./appraisal-data')

/**
 *
 * @module readme-inspector/readme-info
 *
 * Provides information about a README document.
 *
 * @typedef ReadmeInfo
 * @type {object}
 *
 * @prop {AppraisalData} appraisal - Represents the
 *   quantifiable quality of a README file.
 * @prop {Error} err=null - Provides `Error` information
 *   in case of exceptions.
 * @prop {boolean} isPresent - Declares whether a README
 *   document was found.
 * @prop {external:Contents} value - A README `Contents`
 *   instance.
 *
 * @example
 * {
 *   "appraisal":{
 *     "breakdown":{
 *       "cumulativeCodeBlockLength":10,
 *       "hasLists":10,
 *       "lowCodeBlockPenalty":0,
 *       "numberOfCodeBlocks":40,
 *       "numberOfGifs":0,
 *       "numberOfImages":15,
 *       "numberOfNonCodeSections":30
 *     },
 *     "error":null,
 *     "score":100,
 *     "url":null
 *   },
 *   "err":null,
 *   "isPresent":true,
 *   "value":{
 *     "name":"README.md",
 *     "path":"README.md",
 *     "sha":"7a4be917e2f73fe701170753ef31f278936a7499",
 *     "size":37410,
 *     "url":"https://api.github.com/repos/commonality/readme-inspector/contents/README.md?ref=master",
 *     "html_url":"https://github.com/commonality/readme-inspector/blob/master/README.md",
 *     "git_url":"https://api.github.com/repos/commonality/readme-inspector/git/blobs/7a4be917e2f73fe701170753ef31f278936a7499",
 *     "download_url":"https://raw.githubusercontent.com/commonality/readme-inspector/master/README.md",
 *     "type":"file",
 *     "content":"...",
 *     "encoding":"base64",
 *     "_links":{
 *       "self":"https://api.github.com/repos/commonality/readme-inspector/contents/README.md?ref=master",
 *       "git":"https://api.github.com/repos/commonality/readme-inspector/git/blobs/7a4be917e2f73fe701170753ef31f278936a7499",
 *       "html":"https://github.com/commonality/readme-inspector/blob/master/README.md"
 *     }
 *   }
 * }
 */

const readmeInfo = {
  'appraisal': Object.assign({
  }, appraisalData),
  'err': null,
  'isPresent': null,
  'value': null
}

module.exports = readmeInfo
